src/agent/onefuzz/src/syncdir.rs

```diff
@@ -101,18 +107,66 @@ impl SyncedDir {
         }
     }
 
+    // Conditionally upload a report, if it would not be a duplicate.
+    pub async fn upload<T: Serialize>(&self, name: &str, data: &T) -> Result<bool> {
+        match self.url.as_file_path() {
+            Some(path) => {
+                let path = path.join(name);
+                if !exists(&path).await? {
+                    let data = serde_json::to_vec(&data)?;
+                    fs::write(path, data).await?;
+                    Ok(true)
+                } else {
+                    Ok(false)
+                }
+            }
+            None => {
+                let url = self.url.blob(name).url();
+                let blob = BlobClient::new();
+                let result = blob
+                    .put(url.clone())
+                    .json(data)
+                    // Conditional PUT, only if-not-exists.
+                    // https://docs.microsoft.com/en-us/rest/api/storageservices/specifying-conditional-headers-for-blob-service-operations
+                    .header("If-None-Match", "*")
+                    .send_retry_default()
+                    .await
+                    .context("Uploading blob")?;
+
+                Ok(result.status() == StatusCode::CREATED)
+            }
+        }
+    }
+
     async fn file_monitor_event(&self, event: Event) -> Result<()> {
         debug!("monitoring {}", self.path.display());
         let mut monitor = DirectoryMonitor::new(self.path.clone());
         monitor.start()?;
 
-        let mut uploader = self.url.as_ref().map(|x| BlobUploader::new(x.url()));
+        if let Some(path) = self.url.as_file_path() {
+            fs::create_dir_all(&path).await?;
+
+            while let Some(item) = monitor.next().await {
+                event!(event.clone(); EventData::Path = item.display().to_string());
+                let file_name = item
+                    .file_name()
+                    .ok_or_else(|| anyhow!("invalid file path"))?;
+                let destination = path.join(file_name);
+                if let Err(err) = fs::copy(&item, &destination).await {
+                    warn!(
+                        "Couldn't upload file.  source:{:?} destination:{:?} err:{}",
+                        item, destination, err
+                    );
+                }
+            }
+        } else {
+            let mut uploader = BlobUploader::new(self.url.url().clone());
+
+            while let Some(item) = monitor.next().await {
+                event!(event.clone(); EventData::Path = item.display().to_string());
 
-        while let Some(item) = monitor.next().await {
-            event!(event.clone(); EventData::Path = item.display().to_string());
-            if let Some(uploader) = &mut uploader {
                 if let Err(err) = uploader.upload(item.clone()).await {
-                    bail!(
+                    warn!(
```

Comment:
Why did you change this from a failure condition to warn?  I'm concerned this will swallow errors.

Reply:
see my other comment about above. This change was to have the same behavior for local and managed mode.
My thought was that we would not want to stop monitoring the whole directory if one upload fails

Reply:
upload uses the reqwest-retry logic to try to recover.  My thought is if we fail even with automatic recovery, we should hard-fail the task.

Reply:
added a file existence check to decide between warning and bailing